File: src/main/java/in/dragonbra/javasteam/steam/handlers/steammasterserver/SteamMasterServer.java
package in.dragonbra.javasteam.steam.handlers.steammasterserver;

import in.dragonbra.javasteam.base.ClientMsgProtobuf;
import in.dragonbra.javasteam.base.IPacketMsg;
import in.dragonbra.javasteam.enums.EMsg;
import in.dragonbra.javasteam.handlers.ClientMsgHandler;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverGameservers.CMsgClientGMSServerQuery;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverGameservers.CMsgGMSClientServerQueryResponse;
import in.dragonbra.javasteam.steam.handlers.steammasterserver.callback.QueryCallback;
import in.dragonbra.javasteam.types.JobID;
import in.dragonbra.javasteam.util.NetHelpers;
import in.dragonbra.javasteam.util.compat.Consumer;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * This handler is used for requesting server list details from Steam.
 */
public class SteamMasterServer extends ClientMsgHandler {

    private Map<EMsg, Consumer<IPacketMsg>> dispatchMap;

    public SteamMasterServer() {
        dispatchMap = new HashMap<>();

        dispatchMap.put(EMsg.GMSClientServerQueryResponse, this::handleServerQueryResponse);

        dispatchMap = Collections.unmodifiableMap(dispatchMap);
    }

    /**
     * Requests a list of servers from the Steam game master server.
     * Results are returned in a {@link QueryCallback}.
     *
     * @param details The details for the request.
     * @return The Job ID of the request. This can be used to find the appropriate {@link QueryCallback}.
     */
    public JobID serverQuery(QueryDetails details) {
        if (details == null) {
            throw new IllegalArgumentException("details is null");
        }

        ClientMsgProtobuf<CMsgClientGMSServerQuery.Builder> query =
                new ClientMsgProtobuf<>(CMsgClientGMSServerQuery.class, EMsg.ClientGMSServerQuery);
        JobID jobID = client.getNextJobID();
        query.setSourceJobID(jobID);

        query.getBody().setAppId(details.getAppID());

        if (details.getGeoLocatedIP() != null) {
            query.getBody().setGeoLocationIp(NetHelpers.getIPAddress(details.getGeoLocatedIP()));
        }

        query.getBody().setFilterText(details.getFilter());
        query.getBody().setRegionCode(details.getRegion().code());

        query.getBody().setMaxServers(details.getMaxServers());

        client.send(query);

        return jobID;
    }

    @Override
    public void handleMsg(IPacketMsg packetMsg) {
        if (packetMsg == null) {
            throw new IllegalArgumentException("packetMsg is null");
        }

        Consumer<IPacketMsg> dispatcher = dispatchMap.get(packetMsg.getMsgType());
        if (dispatcher != null) {
            dispatcher.accept(packetMsg);
        }
    }

    private void handleServerQueryResponse(IPacketMsg packetMsg) {
        ClientMsgProtobuf<CMsgGMSClientServerQueryResponse.Builder> queryResponse =
                new ClientMsgProtobuf<>(CMsgGMSClientServerQueryResponse.class, packetMsg);

        client.postCallback(new QueryCallback(queryResponse.getTargetJobID(), queryResponse.getBody()));
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steammasterserver/SteamMasterServer.kt
package `in`.dragonbra.javasteam.steam.handlers.steammasterserver

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EMsg
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserverGameservers.CMsgClientGMSServerQuery
import `in`.dragonbra.javasteam.steam.handlers.ClientMsgHandler
import `in`.dragonbra.javasteam.steam.handlers.steammasterserver.callback.QueryCallback
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import `in`.dragonbra.javasteam.types.AsyncJobSingle
import `in`.dragonbra.javasteam.util.NetHelpers

/**
 * This handler is used for requesting server list details from Steam.
 */
class SteamMasterServer : ClientMsgHandler() {

    /**
     * Requests a list of servers from the Steam game master server.
     * Results are returned in a [QueryCallback].
     *
     * @param details The details for the request.
     * @return The Job ID of the request. This can be used to find the appropriate [QueryCallback].
     */
    fun serverQuery(details: QueryDetails): AsyncJobSingle<QueryCallback> {
        val query = ClientMsgProtobuf<CMsgClientGMSServerQuery.Builder>(
            CMsgClientGMSServerQuery::class.java,
            EMsg.ClientGMSServerQuery
        ).apply {
            sourceJobID = client.getNextJobID()

            body.appId = details.appID

            if (details.geoLocatedIP != null) {
                body.geoLocationIp = NetHelpers.getIPAddress(details.geoLocatedIP!!)
            }

            body.filterText = details.filter
            body.regionCode = details.region.code().toInt()

            body.maxServers = details.maxServers
        }

        client.send(query)

        return AsyncJobSingle(this.client, query.sourceJobID)
    }

    /**
     * Handles a client message. This should not be called directly.
     * @param packetMsg The packet message that contains the data.
     */
    override fun handleMsg(packetMsg: IPacketMsg) {
        // ignore messages that we don't have a handler function for
        val callback = getCallback(packetMsg) ?: return

        client.postCallback(callback)
    }

    companion object {
        private fun getCallback(packetMsg: IPacketMsg): CallbackMsg? = when (packetMsg.msgType) {
            EMsg.GMSClientServerQueryResponse -> QueryCallback(packetMsg)
            else -> null
        }
    }
}
